janus_server/src/datastore.rs

```diff
@@ -128,6 +128,15 @@ impl<C: Clock> Datastore<C> {
         tx.tx.commit().await?;
         Ok(rslt)
     }
+
+    /// Attempt to fetch a connection from the connection pool, to check if
+    /// the database is accessible. This will either create a new database
+    /// connection or recycle an existing one, and then return the connection
+    /// back to the pool.
+    pub async fn check_connection_pool(&self) -> Result<(), Error> {
+        let _client = self.pool.get().await?;
```

Comment:
I'm not sure I like the idea that a health check will (potentially) create new connections to the database. This will introduce additional connections, which can be a valuable resource. More importantly, if we ever introduce a maximum number of connections per process (which seems likely), a process that's simply busy might appear unhealthy. It would be taken down, which would lead to other processes taking on this load, and themselves becoming busier -- at worst, this could lead to cascading failure due to (self-inflicted) overload.

If all we want is a health check, IMO the health-check endpoint should just return 200.

If we want an effective "initial liveness" check as well, perhaps the datastore could track whether it has ever succeeded at getting a client, and return true from here if so? (and if it hasn't ever gotten a client, then it would do so here, but only at most once up to concurrency of health-check requests)

But an even better to the initial liveness check might be to retrieve a client once at process startup (ie in the main function somewhere), before we start listening on a port for requests. Then we could just do the just-return-200 in the health check.

Reply:
It's possible that what we want here is a _startup probe_ as well as a health check. IIUC, Kubernetes won't consider a container to be started up (and hence won't add it to the backends of a service) until the startup probe succeeds, but once it does succeed, it won't hit the startup probe path again. So that would be a good place to check that we can connect to the database, that the necessary cryptography keys are available, etc..